Getting a panic after upgrading from `3.88.1` to `3.94.1`

### What happened?

Hi

just upgraded from Pulumi version 3.88.1 to 3.94.1, and am now seeing the following error when trying to preview some stacks:

```
panic: fatal: A failure has occurred: unknown event type 'policy-load'
```

The word `policy-load` appears to be related to changes introduced in https://github.com/pulumi/pulumi/pull/14493

Here is the full panic when running `pulumi preview`.

```
panic: fatal: A failure has occurred: unknown event type 'policy-load'

goroutine 31 [running]:
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
	/private/tmp/pulumi-20231116-4266-1mxjdrm/sdk/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Failf({0x1044977fd?, 0xb?}, {0x14000cb15d0?, 0x10272e024?, 0x107144a68?})
	/private/tmp/pulumi-20231116-4266-1mxjdrm/sdk/go/common/util/contract/fail.go:32 +0xbc
github.com/pulumi/pulumi/pkg/v3/backend/display.RenderDiffEvent({{0x10444b37c?, 0x14000caaaf8?}, {0x104f08e20?, 0x1068f2260?}}, 0x14001d47100?, {{0x104424256, 0x5}, 0x0, 0x0, 0x0, ...})
	/private/tmp/pulumi-20231116-4266-1mxjdrm/pkg/backend/display/diff.go:129 +0x728
github.com/pulumi/pulumi/pkg/v3/backend/display.ShowDiffEvents({0x1400148e750, 0x11}, 0x14001dab020, 0x14001dab080, {{0x104424256, 0x5}, 0x0, 0x0, 0x0, 0x0, ...})
	/private/tmp/pulumi-20231116-4266-1mxjdrm/pkg/backend/display/diff.go:85 +0x4a8
github.com/pulumi/pulumi/pkg/v3/backend/display.ShowEvents({0x1400148e750, 0x11}, {0x10442d95b, 0x7}, {{0x14001bd0c80?, 0x7?}}, {0x14001bd0f90, 0x10}, {0x14001dfe000, 0x61}, ...)
	/private/tmp/pulumi-20231116-4266-1mxjdrm/pkg/backend/display/display.go:87 +0x1f8
created by github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudUpdate).RecordAndDisplayEvents in goroutine 15
	/private/tmp/pulumi-20231116-4266-1mxjdrm/pkg/backend/httpstate/state.go:131 +0x290

```

### Example

Running `pulumi preview` on multiple stacks seems to trigger this problem.

### Output of `pulumi about`

```
CLI          
Version      3.94.1
Go Version   go1.21.4
Go Compiler  gc

Plugins
NAME    VERSION
nodejs  unknown

Host     
OS       darwin
Version  14.0
Arch     arm64

This project is written in nodejs: executable='/Users/redacted/.nvm/versions/node/v18.17.0/bin/node' version='v18.17.0'

Current Stack: xxxxxxxxxx/date-eks-cluster/dev

TYPE                                                        URN
pulumi:pulumi:Stack                                         urn:pulumi:dev::date-eks-cluster::pulumi:pulumi:Stack::date-eks-cluster-dev
pulumi:providers:aws                                        urn:pulumi:dev::date-eks-cluster::pulumi:providers:aws::default_5_42_0
vpc                                                         urn:pulumi:dev::date-eks-cluster::vpc::us-east-1-dev-date
aws:ec2/vpc:Vpc                                             urn:pulumi:dev::date-eks-cluster::vpc$aws:ec2/vpc:Vpc::us-east-1-dev-date
aws:ec2/routeTable:RouteTable                               urn:pulumi:dev::date-eks-cluster::vpc$aws:ec2/vpc:Vpc$aws:ec2/routeTable:RouteTable::us-east-1-dev-date-private-us-east-1a
aws:ec2/subnet:Subnet                                       urn:pulumi:dev::date-eks-cluster::vpc$aws:ec2/vpc:Vpc$aws:ec2/subnet:Subnet::us-east-1-dev-date-private-us-east-1f
[... // content removed //...]
kubernetes:helm.sh/v3:Release                               urn:pulumi:dev::date-eks-cluster::k8s-datadog$kubernetes:helm.sh/v3:Release::datadog


Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/xxxxxxxxxxx
User           xxxxxxxxxxx
Organizations  xxxxxxxxxxx, xxxxxxxxxxx
Token type     personal

Pulumi locates its logs in /var/folders/7m/bwx6v5vs4nd006n02xmpsl680000gq/T/ by default

```

### Additional context

If needed, we have a stack state as a json.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20231117--cli-engine--fix-an-issue-where-the-cli-could-panic-because-of-a-newly-introduced-event.yaml
+++ b/changelog/pending/20231117--cli-engine--fix-an-issue-where-the-cli-could-panic-because-of-a-newly-introduced-event.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: cli/engine
+  description: Fix an issue where the CLI could panic because of a newly introduced event

--- a/pkg/backend/display/diff.go
+++ b/pkg/backend/display/diff.go
@@ -98,6 +98,8 @@ func RenderDiffEvent(event engine.Event, seen map[resource.URN]engine.StepEventM
 	switch event.Type {
 	case engine.CancelEvent:
 		return ""
+	case engine.PolicyLoadEvent:
+		return ""
 
 		// Currently, prelude, summary, and stdout events are printed the same for both the diff and
 		// progress displays.

--- a/pkg/backend/display/events.go
+++ b/pkg/backend/display/events.go
@@ -168,6 +168,9 @@ func ConvertEngineEvent(e engine.Event, showSecrets bool) (apitype.EngineEvent, 
 			Steps:    p.Steps,
 		}
 
+	case engine.PolicyLoadEvent:
+		apiEvent.PolicyLoadEvent = &apitype.PolicyLoadEvent{}
+
 	default:
 		return apiEvent, fmt.Errorf("unknown event type %q", e.Type)
 	}
@@ -373,6 +376,9 @@ func ConvertJSONEvent(apiEvent apitype.EngineEvent) (engine.Event, error) {
 			Status:   resource.Status(p.Status),
 			Steps:    p.Steps,
 		})
+
+	case apiEvent.PolicyLoadEvent != nil:
+		event = engine.NewEvent(engine.PolicyLoadEventPayload{})
 
 	default:
 		return event, errors.New("unknown event type")

--- a/pkg/backend/display/json.go
+++ b/pkg/backend/display/json.go
@@ -203,6 +203,9 @@ func ShowPreviewDigest(events <-chan engine.Event, done chan<- bool, opts Option
 		case engine.PolicyViolationEvent:
 			// At this point in time, we don't handle policy events in JSON serialization
 			continue
+		case engine.PolicyLoadEvent:
+			// At this point in time, we don't handle policy events in JSON serialization
+			continue
 		case engine.SummaryEvent:
 			// At the end of the preview, a summary event indicates the final conclusions.
 			p := e.Payload().(engine.SummaryEventPayload)

--- a/pkg/backend/display/query.go
+++ b/pkg/backend/display/query.go
@@ -93,6 +93,9 @@ func renderQueryEvent(event engine.Event, opts Options) string {
 		contract.Failf("query mode does not support resource operations")
 		return ""
 
+	case engine.PolicyLoadEvent:
+		return ""
+
 	default:
 		contract.Failf("unknown event type '%s'", event.Type)
 		return ""

--- a/pkg/backend/display/watch.go
+++ b/pkg/backend/display/watch.go
@@ -46,7 +46,7 @@ func ShowWatchEvents(op string, events <-chan engine.Event, done chan<- bool, op
 		// For all other events, use the payload to build up the JSON digest we'll emit later.
 		switch e.Type {
 		// Events occurring early:
-		case engine.PreludeEvent, engine.SummaryEvent, engine.StdoutColorEvent:
+		case engine.PreludeEvent, engine.SummaryEvent, engine.StdoutColorEvent, engine.PolicyLoadEvent:
 			// Ignore it
 			continue
 		case engine.PolicyViolationEvent:

--- a/sdk/go/common/apitype/events.go
+++ b/sdk/go/common/apitype/events.go
@@ -195,6 +195,9 @@ type ResOpFailedEvent struct {
 	Steps    int               `json:"steps"`
 }
 
+// PolicyLoadEvent is emitted when a policy starts loading
+type PolicyLoadEvent struct{}
+
 // EngineEvent describes a Pulumi engine event, such as a change to a resource or diagnostic
 // message. EngineEvent is a discriminated union of all possible event types, and exactly one
 // field will be non-nil.
@@ -221,6 +224,7 @@ type EngineEvent struct {
 	ResOpFailedEvent       *ResOpFailedEvent       `json:"resOpFailedEvent,omitempty"`
 	PolicyEvent            *PolicyEvent            `json:"policyEvent,omitempty"`
 	PolicyRemediationEvent *PolicyRemediationEvent `json:"policyRemediationEvent,omitempty"`
+	PolicyLoadEvent        *PolicyLoadEvent        `json:"policyLoadEvent,omitempty"`
 }
 
 // EngineEventBatch is a group of engine events.

